Don't run resource type check if within check interval

clarafu opened the pull request:
## Changes proposed by this PR

closes #8244

We seemed to be running a check for resource types for every build because we didn't have anything within the `WaitToRun` code for checking if the check interval had elapsed.

This fix will help reduce the number of resource type checks, since it is not really necessary to run a resource type check for every build.

## Notes to reviewer

I'm still thinking about whether we want to make resource types get force checked if you submit a manual build. I think initially when this was implemented, we decided not to in order to not spam a ton of resource type checks if the user creates a lot of manual builds. Especially if a lot of resources use the same resource type within the job. I am leaning towards keeping it as is, where we don't force resource type checks on manual builds. The way you can force a resource type check is by running `fly check-resource` which can recursively check the resource types or `fly check-resource-type`.

## Release Note

* Fixes a bug where resource types were getting checked every build. Now they will respect the resource checking interval and not run a check if the interval has not elapsed.

clarafu committed:
don't run resource type check if within check interval

We seemed to be running a check for resource types for every build
because we didn't have anything within the WaitToRun code for checking
if the check interval had elapsed.

This fix will help reduce the number of resource type checks, since it
is not really necessary to run a resource type check for every build.

[#8244]

Signed-off-by: Clara Fu <fclara@vmware.com>

evanchaoli commented:
LGTM!!! This is exactly what I am going to submit.

xtremerui reviewed (commented):
There are failures in both testflight and watsjs could you take a look?

clarafu commented on atc/engine/check_delegate_test.go:
```diff
@@ -440,6 +440,37 @@ var _ = Describe("CheckDelegate", func() {
+					fakeResourceConfigScope.LastCheckReturns(db.LastCheck{
+						StartTime: now,
+						EndTime:   now,
+						Succeeded: true,
```
Oh I actually made it so that the last check end time is before the build start time so that it won't return false because of the first part of the `OR`. But if you think adding this is better than I will add it!

xtremerui replied:
I see. Guess I just want to make sure which taks priority between last check status and elapsed time in the `if` clause. In above test i see it test with last check status false without considerate elapse time interval.

https://github.com/concourse/concourse/blob/c6e9838023a40599d41d7d1a7aabf057ead03d71/atc/engine/check_delegate_test.go#L400-L412

clarafu committed:
run resource type check if manual check

Added extra logic to the WaitToRun so that it will run the check if it
is a manual check (Skipinterval is true) even if the interval has no
elapsed since the last check.

[#8244]

Signed-off-by: Clara Fu <fclara@vmware.com>

clarafu committed:
always check if last check is unsuccessful

we should always run the check if the last check was unsuccessful.

[#8244]

Signed-off-by: Clara Fu <fclara@vmware.com>

xtremerui reviewed (approved):
Those new test cases make me feel rest assured. Thx for the PR!

xtremerui commented:
> we don't force resource type checks on manual builds.

This change is also reflecting on image check for task, the topgun test failure could be found here https://ci.concourse-ci.org/teams/main/pipelines/concourse/jobs/bosh-topgun-both/builds/1138#L6253b9d8:4726:4739

<details>
<p>

```
STEP: [1650996802.141039371] running: /tmp/gexec_artifacts221771373/g1793001697/fly -t concourse-topgun-both-8 --print-table-headers containers
handle                                worker                                pipeline  job      build #  build id  type   name     attempt
02be3bd8-800f-4486-4eb3-6a86366063ec  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  1        1         check  image    n/a    
11fd86c3-bbba-464f-4a19-1fede6127ab4  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  1        1         get    image    n/a    
20d47d05-6f9b-4b90-7d93-d49902f7f1a9  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      none     check    2         check  my-time  n/a    
23df2f8d-f4f0-43b3-59ff-4b4e4908e140  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  1        1         get    10m      n/a    
9e14dc64-208b-4666-4132-e849f75b2a6f  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  1        1         task   fail     n/a    
c804111b-faf2-4530-5cb0-4c4c49780ff5  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      none     check    2         check  10m      n/a    
cbb69acf-fbd1-49f1-61ff-d6372bfc24eb  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      none     check    2         get    my-time  n/a    
STEP: triggering the build again
STEP: [1650996802.188416481] running: /tmp/gexec_artifacts221771373/g1793001697/fly -t concourse-topgun-both-8 trigger-job -w -j pipe/get-10m
started pipe/get-10m #2

initializing check: my-time
selected worker: 5fa5eea7-3b25-4a80-bcd7-468d0876ec87
INFO: found existing resource cache

selected worker: 5fa5eea7-3b25-4a80-bcd7-468d0876ec87
INFO: found existing resource cache

initializing
initializing check: image
selected worker: 5fa5eea7-3b25-4a80-bcd7-468d0876ec87
INFO: found existing resource cache

selected worker: 5fa5eea7-3b25-4a80-bcd7-468d0876ec87
running false
failed
STEP: expecting only one additional check container for the task's image check
STEP: [1650996805.733932257] running: /tmp/gexec_artifacts221771373/g1793001697/fly -t concourse-topgun-both-8 --print-table-headers containers
handle                                worker                                pipeline  job      build #  build id  type   name     attempt
02be3bd8-800f-4486-4eb3-6a86366063ec  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  1        1         check  image    n/a    
11fd86c3-bbba-464f-4a19-1fede6127ab4  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  1        1         get    image    n/a    
20d47d05-6f9b-4b90-7d93-d49902f7f1a9  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      none     check    2         check  my-time  n/a    
23df2f8d-f4f0-43b3-59ff-4b4e4908e140  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  1        1         get    10m      n/a    
4c88d8c9-b677-4752-7fce-08d89cc4b5ce  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  2        3         task   fail     n/a    
9e14dc64-208b-4666-4132-e849f75b2a6f  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      get-10m  1        1         task   fail     n/a    
c804111b-faf2-4530-5cb0-4c4c49780ff5  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      none     check    2         check  10m      n/a    
cbb69acf-fbd1-49f1-61ff-d6372bfc24eb  5fa5eea7-3b25-4a80-bcd7-468d0876ec87  pipe      none     check    2         get    my-time  n/a 
```

</p></details>

As shown above, no new check container was created after the job was triggered again, since the check interval had not elapsed yet. The problem here is the image resource is anonymously defined in task yaml, AFAIK there is no way to force check this type of image resource, unlike resource type.

I think the behaviour of not forcing check should not inlude above kind of image resource check.
